### PR TITLE
Architecture doc for explaining the game engine and agent architectures

### PR DESCRIPTION
**Pull request by sujayakar**
I'll finish writing up the agent bit later tonight.

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+### Simulation
+
+Other than when processing player inputs, the game state can change over time in the background as the
+simulation runs time forward. For example, if a player has decided to move along a path, their position
+will gradually update as time moves forward. Similarly, if two players collide into each other, they'll
+notice and replan their paths, trying to avoid obstacles.
+
+### Message data model
+
+We manage the tables for tracking chat messages in separate tables not affiliated with the game engine.
```
Comment (ianmacartney) [inline ARCHITECTURE.md:81]: Is it easy to explain why? Or what heuristic folks might want to use to decide when to put things in the game state vs. outside?

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+  when possible, and store game state in regular tables.
+- Be as similar to existing engines as possible, so it's easy to change the behavior. We chose a
+  `tick()` based model for simulation since it's commonly used elsewhere and intuitive.
+- Decouple agent behavior from the game engine. It's nice to allow human players and AI agents to do
+  all the same things in the game.
+
+These design goals imply some inherent limitations:
+
+- All data is loaded into memory each step. The active game state loaded by the game should be small
+  enough to fit into and load and save frequently. Try to keep game state to less than a few dozen
```
Comment (ianmacartney) [inline ARCHITECTURE.md:276]: fit into *memory*

**Review by ianmacartney — COMMENTED** [review]
very good. concise and reasonably comprehensive. There's a lot to cover, and not everyone will follow everything (in part it requires a background understanding of `tick` etc.

There's some tactical bits I'm not sure if we should include, like "if one aspect needs higher latency requirements, you can have it in a table outside of game state, and just log key bits of it in the game state, like we do for messages" 

But I think the really nuanced bits are probably better sorted out in Discord ad-hoc.

File: ARCHITECTURE.md (hunk 1 above)
Comment (sujayakar) [reply ARCHITECTURE.md:81]: updated with an explanation! and yeah I included your tactical bit from above in the overview with a pointer to this section.

**Pull request merged**